diff --git a/vendor/github.com/circonus-labs/circonus-gometrics/checkmgr/checkmgr.go b/vendor/github.com/circonus-labs/circonus-gometrics/checkmgr/checkmgr.go
index 1237ba1..99dccf9 100644
--- a/vendor/github.com/circonus-labs/circonus-gometrics/checkmgr/checkmgr.go
+++ b/vendor/github.com/circonus-labs/circonus-gometrics/checkmgr/checkmgr.go
@@ -249,7 +249,7 @@ func NewCheckManager(cfg *Config) (*CheckManager, error) {
 		hn = "unknown"
 	}
 	if cm.checkInstanceID == "" {
-		cm.checkInstanceID = CheckInstanceIDType(fmt.Sprintf("%s:%s", hn, an))
+		cm.checkInstanceID = CheckInstanceIDType(fmt.Sprintf("%s:%s", hn, cfg.API.TokenApp))
 	}
 
 	if cm.checkSearchTag == "" {
